sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import javax.annotation.Nullable
+
+import scala.annotation.tailrec
+
+import org.apache.spark.sql.catalyst.analysis.TypeCoercion.hasStringType
+import org.apache.spark.sql.catalyst.expressions.{BinaryExpression, Cast, Collate, ComplexTypeMergingExpression, CreateArray, Elt, ExpectsInputTypes, Expression, Predicate, SortOrder}
+import org.apache.spark.sql.errors.QueryCompilationErrors
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{AbstractDataType, ArrayType, DataType, StringType}
+
+object CollationTypeCasts extends TypeCoercionRule {
+  override val transform: PartialFunction[Expression, Expression] = {
+    case e if !e.childrenResolved => e
+    case sc @ (_: BinaryExpression
+               | _: ComplexTypeMergingExpression
+               | _: CreateArray
+               | _: Elt
+               | _: ExpectsInputTypes
+               | _: Predicate
+               | _: SortOrder) =>
+      val newChildren = collateToSingleType(sc.children)
+      sc.withNewChildren(newChildren)
+  }
+
+  /**
+   * Extracts StringTypes from filtered hasStringType
+   */
+  @tailrec
+  private def extractStringType(dt: DataType): StringType = dt match {
+    case st: StringType => st
+    case ArrayType(et, _) => extractStringType(et)
+  }
+
+  /**
+   * Casts given expression to collated StringType with id equal to collationId only
+   * if expression has StringType in the first place.
+   * @param expr
+   * @param collationId
+   * @return
+   */
+  def castStringType(expr: Expression, st: StringType): Option[Expression] =
+    castStringType(expr.dataType, st).map { dt => Cast(expr, dt)}
+
+  private def castStringType(inType: AbstractDataType, castType: StringType): Option[DataType] = {
+    @Nullable val ret: DataType = inType match {
+      case st: StringType if st.collationId != castType.collationId => castType
+      case ArrayType(arrType, nullable) =>
```

Comment:
I disagree with special-case array type. The code looks broken. It assumes the children of the given expression can have both string type and array of string type, then tries to find a common collation between the string type child and the array element. This makes no sense without knowing the semantic of the given expression.

Reply:
A simple example is ConcatWs. It can have ArrayType(StringType, _) for input strings and StringType for separator as parameters. What collations do we want for this then? We need to cast the ArrayType into a proper collation if separator is explicit.

Reply:
Then please match the ConcatWs expression explicitly to handle this case. What I disagree with is to do this blindly for all expressions.